x-pack/plugins/ml/public/application/jobs/new_job/pages/components/common/datafeed_preview_flyout/datafeed_preview.tsx

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { FC, useState, useEffect, useMemo } from 'react';
+import { i18n } from '@kbn/i18n';
+import { FormattedMessage } from '@kbn/i18n/react';
+import {
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiTitle,
+  EuiSpacer,
+  EuiLoadingSpinner,
+  EuiButton,
+} from '@elastic/eui';
+
+import { CombinedJob } from '../../../../../../../../common/types/anomaly_detection_jobs';
+import { MLJobEditor } from '../../../../../jobs_list/components/ml_job_editor';
+import { mlJobService } from '../../../../../../services/job_service';
+import { ML_DATA_PREVIEW_COUNT } from '../../../../../../../../common/util/job_utils';
+
+export const DatafeedPreview: FC<{
+  combinedJob: CombinedJob | null;
+  heightOffset?: number;
+}> = ({ combinedJob, heightOffset = 0 }) => {
+  // the ace editor requires a fixed height
+  const editorHeight = useMemo(() => `${window.innerHeight - 230 - heightOffset}px`, [
+    heightOffset,
+  ]);
+  const [loading, setLoading] = useState(false);
+  const [previewJsonString, setPreviewJsonString] = useState('');
+  const [outOfDate, setOutOfDate] = useState(false);
+  const [combinedJobString, setCombinedJobString] = useState('');
+
+  useEffect(() => {
+    try {
+      if (combinedJob !== null) {
+        if (combinedJobString === '') {
+          // first time, set the string and load the preview
+          loadDataPreview();
+        } else {
+          setOutOfDate(JSON.stringify(combinedJob) !== combinedJobString);
+        }
+      }
+    } catch (error) {
+      // fail silently
+    }
+  }, [combinedJob]);
+
+  async function loadDataPreview() {
```

Comment:
nit: function is reinitialized on each render

Reply:
Updated in https://github.com/elastic/kibana/pull/72117/commits/0cbdfc9e16b2873205b72590b80585eb3c0c739c